firmware/thruna/FSM/Src/Io/Io_AcceleratorPedals.c

```diff
@@ -1,76 +1,127 @@
 
-#define MIN_PEDAL_VOLTAGE (0.5f)
-#define MAX_PEDAL_VOLTAGE (3.1f)
+#define SAPPS_LEN_PER_VOLTAGE ((SAPPS_LENGTH_FULL_PRESSED_MM - SAPPS_LENGTH_UNPRESSED_MM) / (SAPPS_FULL_PRESSED_POT_V - SAPPS_UNPRESSED_POT_V))
+// PAPPS length offset in mm
+#define SAPPS_LEN_OFFSET_MM (SAPPS_LENGTH_UNPRESSED_MM - SAPPS_LEN_PER_VOLTAGE * SAPPS_UNPRESSED_POT_V) // Yint calculated based on the measured unpressed length resulting in the unpressed pot val
+#define SAPPS_RAW_VOLTAGE_TO_LEN_MM(voltage) (voltage * SAPPS_LEN_PER_VOLTAGE + SAPPS_LEN_OFFSET_MM)
+
+#define PAPPS_MIN_V (PAPPS_UNPRESSED_POT_V - 0.5f)
+#define PAPPS_MAX_V (PAPPS_FULL_PRESSED_POT_V + 0.5f)
+
+#define SAPPS_MIN_V (SAPPS_UNPRESSED_POT_V - 0.5f)
+#define SAPPS_MAX_V (SAPPS_FULL_PRESSED_POT_V + 0.5f)
 
-// NONLINEAR APPROXIMATION
-#define FULLY_UNPRESSED_POT_LENGTH_MM (212.2f)
-#define FULLY_PRESSED_POT_LENGTH_MM (165.6f)
+#define DEAD_ZONE_PERCENT (10.0f)
 
-#define PAPPS_DEADZONE_PERCENT (10.0f)
-#define SAPPS_DEADZONE_PERCENT (8.0f)
+
+// function created to calculate angles relative to rest angle (since we dont want a negative angle we will do rest pos - current pos)
+float Io_AcceleratorPedals_Angle_calc(float cos_law_coefficent, float cos_law_denominator, float pot_len, float max_len)
+{
+    float angle = (acosf(cos_law_coefficent - (powf(max_len, 2) / cos_law_denominator)))  -
+                  (acosf(cos_law_coefficent - (powf(pot_len, 2) / cos_law_denominator)));
+
+    return angle;
+}
 
 float Io_AcceleratorPedals_GetPapps(void)
 {
     // Length calculated via voltage
     float       pedal_voltage = Io_Adc_GetChannelVoltage(ADC1_CHANNEL_2);
     const float pot_len_mm    = PAPPS_RAW_VOLTAGE_TO_LEN_MM(pedal_voltage);
 
-    // Calculate pedal percentage
-    float pedal_percentage = (pot_len_mm - FULLY_UNPRESSED_POT_LENGTH_MM) /
-                                 (FULLY_PRESSED_POT_LENGTH_MM - FULLY_UNPRESSED_POT_LENGTH_MM) * 100.0f -
-                             PAPPS_DEADZONE_PERCENT;
+    float pedal_angle = Io_AcceleratorPedals_Angle_calc(
+        PAPPS_COS_LAW_COEFFICIENT, PAPPS_COS_LAW_DENOMINATOR, pot_len_mm, PAPPS_LENGTH_UNPRESSED_MM);
+    const float max_angle = Io_AcceleratorPedals_Angle_calc(
+        PAPPS_COS_LAW_COEFFICIENT, PAPPS_COS_LAW_DENOMINATOR, PAPPS_LENGTH_FULL_PRESSED_MM, PAPPS_LENGTH_UNPRESSED_MM);
 
-    return CLAMP(pedal_percentage, 0.0f, 100.0f);
-}
+    float pedal_percentage_raw = (pedal_angle / max_angle) * 100.0f;
+
+    if(pedal_percentage_raw <= DEAD_ZONE_PERCENT){
+        return 0;
+    }
+
+    float pedal_percentage = CLAMP(((100 / (100 - DEAD_ZONE_PERCENT)) * (pedal_percentage_raw - DEAD_ZONE_PERCENT)), 0.0f, 100.0f);
+    
+    return pedal_percentage;
+ }
 
-bool Io_AcceleratorPedals_PappsOCSC(void)
+
+bool Io_AcceleratorPedals_PappsOCSC(
+    void) // returns false if within range (no OC or SC) and true if it is outside the range (OS or SC is present)
```

Comment:
Add comments like this to the `.h` file. Look through other `.h` files for an example of how we document functions.